Based on Fedora's icu4c-4_6_1-crossbuild.patch and
icu4c-4_6_1-use-correct-genccode-assembly-on-win32-and-win64.patch.

--- origsrc/icu/source/config/Makefile.inc.in	2013-12-02 19:05:20.001863600 -0600
+++ src/icu/source/config/Makefile.inc.in	2013-12-02 18:58:03.000000000 -0600
@@ -41,7 +41,7 @@ UNICODE_VERSION = @UNICODE_VERSION@
 # The prefix for ICU libraries, normally 'icu'
 ICUPREFIX = icu
 PACKAGE = @PACKAGE@
-LIBICU = lib$(ICUPREFIX)
+LIBICU = $(ICUPREFIX)
 
 # Static library prefix and file extension
 STATIC_PREFIX = s
@@ -95,7 +95,6 @@ LAYOUTEX_STUBNAME = lx
 IO_STUBNAME = io
 TOOLUTIL_STUBNAME = tu
 CTESTFW_STUBNAME = test
- 
 
 
 ### To link your application with ICU:
@@ -106,18 +105,19 @@ CTESTFW_STUBNAME = test
 #    - $(ICULIBS_LAYOUT)  - ICU layout library. 
 #    - $(ICULIBS_ICUIO)   - ICU stdio equivalent library
 
-ICULIBS_COMMON = -l$(ICUPREFIX)uc$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_DATA   = -l$(ICUPREFIX)$(DATA_STUBNAME)$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_I18N   = -l$(ICUPREFIX)$(I18N_STUBNAME)$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_TOOLUTIL = -l$(ICUPREFIX)tu$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_CTESTFW = -l$(ICUPREFIX)ctestfw$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_ICUIO = -l$(ICUPREFIX)io$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_OBSOLETE = -l$(ICUPREFIX)obsolete$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_LAYOUT = -l$(ICUPREFIX)le$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
-ICULIBS_LAYOUTEX = -l$(ICUPREFIX)lx$(ICULIBSUFFIX)$(ICULIBSUFFIX_VERSION)
+ICULIBS_COMMON = -l$(ICUPREFIX)uc$(ICULIBSUFFIX)
+ICULIBS_DATA   = -l$(ICUPREFIX)$(DATA_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_I18N   = -l$(ICUPREFIX)$(I18N_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_TOOLUTIL = -l$(ICUPREFIX)tu$(ICULIBSUFFIX)
+ICULIBS_CTESTFW = -l$(ICUPREFIX)ctestfw$(ICULIBSUFFIX)
+ICULIBS_ICUIO = -l$(ICUPREFIX)io$(ICULIBSUFFIX)
+ICULIBS_OBSOLETE = -l$(ICUPREFIX)obsolete$(ICULIBSUFFIX)
+ICULIBS_LAYOUT = -l$(ICUPREFIX)le$(ICULIBSUFFIX)
+ICULIBS_LAYOUTEX = -l$(ICUPREFIX)lx$(ICULIBSUFFIX)
 ICULIBS_BASE = $(LIBS) -L$(libdir)
 
 # for icu-config to test with
+SO = dll
 ICULIBS_COMMON_LIB_NAME = ${LIBICU}${COMMON_STUBNAME}${ICULIBSUFFIX}${ICULIBSUFFIX_VERSION}.${SO}
 ICULIBS_COMMON_LIB_NAME_A = ${LIBICU}${COMMON_STUBNAME}${ICULIBSUFFIX}.${A}
 
--- origsrc/icu/source/config/icu-config-bottom	2013-12-02 19:05:20.002863600 -0600
+++ src/icu/source/config/icu-config-bottom	2013-12-02 18:58:03.000000000 -0600
@@ -3,7 +3,7 @@
 ## Copyright (c) 2002-2011, International Business Machines Corporation and
 ## others. All Rights Reserved.
 
-ICUUC_FILE="${libdir}/${ICULIBS_COMMON_LIB_NAME}"
+ICUUC_FILE="${bindir}/${ICULIBS_COMMON_LIB_NAME}"
 ICUUC_FILE_A="${libdir}/${ICULIBS_COMMON_LIB_NAME_A}"
     
 #  echo ENABLE RPATH $ENABLE_RPATH and RPATHLDFLAGS=${RPATH_LDFLAGS}
--- origsrc/icu/source/config/icu.pc.in	2013-12-02 19:05:20.004863800 -0600
+++ src/icu/source/config/icu.pc.in	2013-12-02 18:58:03.000000000 -0600
@@ -1,27 +1,27 @@
 # Copyright (C) 2010, International Business Machines Corporation. All Rights Reserved.
 
-prefix = @prefix@
-exec_prefix = @exec_prefix@
-#bindir = @bindir@
-libdir = @libdir@
-includedir = @includedir@
-baselibs = @LIBS@
-#datarootdir = @datarootdir@
-#datadir = @datadir@
-#sbindir = @sbindir@
-#mandir = @mandir@
-#sysconfdir = @sysconfdir@
-CFLAGS = @CFLAGS@ 
-#CXXFLAGS = @CXXFLAGS@ 
-DEFS = @DEFS@ 
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+#bindir=@bindir@
+libdir=@libdir@
+includedir=@includedir@
+baselibs=@LIBS@
+#datarootdir=@datarootdir@
+#datadir=@datadir@
+#sbindir=@sbindir@
+#mandir=@mandir@
+#sysconfdir=@sysconfdir@
+CFLAGS=@CFLAGS@ 
+#CXXFLAGS=@CXXFLAGS@ 
+DEFS=@DEFS@ 
 UNICODE_VERSION=@UNICODE_VERSION@
 ICUPREFIX=icu
 ICULIBSUFFIX=@ICULIBSUFFIX@
-LIBICU=lib${ICUPREFIX}
+LIBICU=${ICUPREFIX}
 #SHAREDLIBCFLAGS=-fPIC
 pkglibdir=${libdir}/@PACKAGE@${ICULIBSUFFIX}/@VERSION@
 #pkgdatadir=${datadir}/@PACKAGE@${ICULIBSUFFIX}/@VERSION@
-ICUDATA_NAME = icudt@LIB_VERSION_MAJOR@@ICUDATA_CHAR@
+ICUDATA_NAME=icudata@LIB_VERSION_MAJOR@@ICUDATA_CHAR@
 #ICUPKGDATA_DIR=@libdir@
 #ICUDATA_DIR=${pkgdatadir}
 ICUDESC=International Components for Unicode
--- origsrc/icu/source/config/mh-mingw	2013-12-02 19:05:28.876371200 -0600
+++ src/icu/source/config/mh-mingw	2013-12-02 18:58:03.000000000 -0600
@@ -44,8 +44,8 @@ LINK.c=       $(CXX) $(CXXFLAGS) $(LDFLA
 LD_SOOPTIONS= -Wl,-Bsymbolic
 
 ## Commands to make a shared library
-SHLIB.c=	$(CC) $(CFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
-SHLIB.cc=	$(CXX) $(CXXFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
+SHLIB.c=	$(CC) $(CFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(LIBPREFIX)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
+SHLIB.cc=	$(CXX) $(CXXFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(LIBPREFIX)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
 
 ## Compiler switch to embed a runtime search path
 LD_RPATH=	
@@ -70,15 +70,11 @@ LIBSICU = $(LIBPREFIX)$(STATIC_PREFIX)$(
 A = a
 
 ## An import library is needed for z/OS and MSVC
-IMPORT_LIB_EXT = .lib
+IMPORT_LIB_EXT = .dll.a
 
-LIBPREFIX=
+#LIBPREFIX=
 
-# Change the stubnames so that poorly working FAT disks and installation programs can work.
-# This is also for backwards compatibility.
-DATA_STUBNAME = dt
-I18N_STUBNAME = in
-LIBICU = $(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)
+LIBICU = $(ICUPREFIX)
 
 # The #M# is used to delete lines for icu-config
 # Current full path directory.
@@ -118,7 +114,7 @@ CURR_SRCCODE_FULL_DIR=$(subst \,/,$(shel
 FINAL_SO_TARGET=$(basename $(SO_TARGET))$(SO_TARGET_VERSION_MAJOR).$(SO)
 MIDDLE_SO_TARGET=$(FINAL_SO_TARGET)
 
-FINAL_IMPORT_LIB = $(dir $(SO_TARGET))$(notdir $(basename $(SO_TARGET)))$(IMPORT_LIB_EXT)#M#
+FINAL_IMPORT_LIB = $(dir $(SO_TARGET))$(LIBPREFIX)$(notdir $(basename $(SO_TARGET)))$(IMPORT_LIB_EXT)#M#
 IMPORT_LIB = $(FINAL_IMPORT_LIB)#M#
 MIDDLE_IMPORT_LIB = $(FINAL_IMPORT_LIB)#M#
 
@@ -127,6 +123,9 @@ PKGDATA_VERSIONING = -r $(SO_TARGET_VERS
 #ICUPKGDATA_INSTALL_DIR = $(shell cygpath -dma $(DESTDIR)$(ICUPKGDATA_DIR))#M#
 #ICUPKGDATA_INSTALL_LIBDIR = $(shell cygpath -dma $(DESTDIR)$(libdir))#M#
 
+# The following is for Makefile.inc's use.
+ICULIBSUFFIX_VERSION =
+
 ## Versioned libraries rules
 #%$(SO_TARGET_VERSION_MAJOR).$(SO): %$(SO_TARGET_VERSION).$(SO)
 #	$(RM) $@ && cp ${<F} $@
@@ -136,11 +135,20 @@ PKGDATA_VERSIONING = -r $(SO_TARGET_VERS
 # Environment variable to set a runtime search path
 LDLIBRARYPATH_ENVVAR = PATH
 
-# The type of assembly to write for generating an object file
-GENCCODE_ASSEMBLY=-a gcc-cygwin
-
 # These are needed to allow the pkgdata GNU make files to work
 PKGDATA_DEFS = -DU_MAKE=\"$(MAKE)\"
 
+# Just the libs.
+ICULIBS_DT	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(DATA_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_UC	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(COMMON_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_I18N	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(I18N_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_LE	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(LAYOUT_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_LX	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(LAYOUTEX_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_IO	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(IO_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_CTESTFW	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(CTESTFW_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_TOOLUTIL = -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(TOOLUTIL_STUBNAME)$(ICULIBSUFFIX)
+# Link commands to link to ICU libs
+LLIBDIR		= -L$(LIBDIR)
+
 ## End Cygwin/MinGW specific setup
 
--- origsrc/icu/source/config/mh-mingw64	2013-12-02 19:05:40.789052500 -0600
+++ src/icu/source/config/mh-mingw64	2013-12-02 19:04:36.436371800 -0600
@@ -44,8 +44,8 @@ LINK.c=       $(CXX) $(CXXFLAGS) $(LDFLA
 LD_SOOPTIONS= -Wl,-Bsymbolic
 
 ## Commands to make a shared library
-SHLIB.c=	$(CC) $(CFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
-SHLIB.cc=	$(CXX) $(CXXFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
+SHLIB.c=	$(CC) $(CFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(LIBPREFIX)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
+SHLIB.cc=	$(CXX) $(CXXFLAGS) $(LDFLAGS) -shared $(LD_SOOPTIONS) -Wl,--enable-auto-import -Wl,--out-implib=$(dir $@)$(LIBPREFIX)$(notdir $(@:$(SO_TARGET_VERSION_MAJOR).$(SO)=))$(IMPORT_LIB_EXT)#M#
 
 ## Compiler switch to embed a runtime search path
 LD_RPATH=	
@@ -70,15 +70,11 @@ LIBSICU = $(LIBPREFIX)$(STATIC_PREFIX)$(
 A = a
 
 ## An import library is needed for z/OS and MSVC
-IMPORT_LIB_EXT = .lib
+IMPORT_LIB_EXT = .dll.a
 
-LIBPREFIX=
+#LIBPREFIX=
 
-# Change the stubnames so that poorly working FAT disks and installation programs can work.
-# This is also for backwards compatibility.
-DATA_STUBNAME = dt
-I18N_STUBNAME = in
-LIBICU = $(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)
+LIBICU = $(ICUPREFIX)
 
 # The #M# is used to delete lines for icu-config
 # Current full path directory.
@@ -118,7 +114,7 @@ CURR_SRCCODE_FULL_DIR=$(subst \,/,$(shel
 FINAL_SO_TARGET=$(basename $(SO_TARGET))$(SO_TARGET_VERSION_MAJOR).$(SO)
 MIDDLE_SO_TARGET=$(FINAL_SO_TARGET)
 
-FINAL_IMPORT_LIB = $(dir $(SO_TARGET))$(notdir $(basename $(SO_TARGET)))$(IMPORT_LIB_EXT)#M#
+FINAL_IMPORT_LIB = $(dir $(SO_TARGET))$(LIBPREFIX)$(notdir $(basename $(SO_TARGET)))$(IMPORT_LIB_EXT)#M#
 IMPORT_LIB = $(FINAL_IMPORT_LIB)#M#
 MIDDLE_IMPORT_LIB = $(FINAL_IMPORT_LIB)#M#
 
@@ -127,6 +123,9 @@ PKGDATA_VERSIONING = -r $(SO_TARGET_VERS
 #ICUPKGDATA_INSTALL_DIR = $(shell cygpath -dma $(DESTDIR)$(ICUPKGDATA_DIR))#M#
 #ICUPKGDATA_INSTALL_LIBDIR = $(shell cygpath -dma $(DESTDIR)$(libdir))#M#
 
+# The following is for Makefile.inc's use.
+ICULIBSUFFIX_VERSION =
+
 ## Versioned libraries rules
 #%$(SO_TARGET_VERSION_MAJOR).$(SO): %$(SO_TARGET_VERSION).$(SO)
 #	$(RM) $@ && cp ${<F} $@
@@ -136,11 +135,20 @@ PKGDATA_VERSIONING = -r $(SO_TARGET_VERS
 # Environment variable to set a runtime search path
 LDLIBRARYPATH_ENVVAR = PATH
 
-# The type of assembly to write for generating an object file
-GENCCODE_ASSEMBLY=-a gcc-mingw64
-
 # These are needed to allow the pkgdata GNU make files to work
 PKGDATA_DEFS = -DU_MAKE=\"$(MAKE)\"
 
+# Just the libs.
+ICULIBS_DT	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(DATA_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_UC	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(COMMON_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_I18N	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(I18N_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_LE	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(LAYOUT_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_LX	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(LAYOUTEX_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_IO	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(IO_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_CTESTFW	= -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(CTESTFW_STUBNAME)$(ICULIBSUFFIX)
+ICULIBS_TOOLUTIL = -l$(STATIC_PREFIX_WHEN_USED)$(ICUPREFIX)$(TOOLUTIL_STUBNAME)$(ICULIBSUFFIX)
+# Link commands to link to ICU libs
+LLIBDIR		= -L$(LIBDIR)
+
 ## End Cygwin64/MinGW64 specific setup
 
--- origsrc/icu/source/configure.in	2013-12-02 19:05:28.863370400 -0600
+++ src/icu/source/configure.in	2013-12-02 18:58:03.000000000 -0600
@@ -598,6 +598,12 @@ sparc-*-solaris*)
 ia64-*-hpux*)
     GENCCODE_ASSEMBLY="-a aCC-ia64"
     ;;
+i686-*-mingw32)
+    GENCCODE_ASSEMBLY="-a gcc-mingw-x86"
+    ;;
+x86_64-*-mingw32)
+    GENCCODE_ASSEMBLY="-a gcc-mingw-x64"
+    ;;
 esac
 AC_SUBST(GENCCODE_ASSEMBLY)
 
--- origsrc/icu/source/tools/toolutil/pkg_genc.c	2013-12-02 19:05:20.008864000 -0600
+++ src/icu/source/tools/toolutil/pkg_genc.c	2013-12-02 18:58:03.000000000 -0600
@@ -149,6 +149,22 @@ static const struct AssemblyType {
 
         ".long ","",HEX_0X
     },
+    {"gcc-mingw-x86",
+        ".globl _%s\n"
+        "\t.section .rdata,\"dr\"\n"
+        "\t.align 4\n"
+        "_%s:\n\n",
+
+        ".long ","",HEX_0X
+    },
+    {"gcc-mingw-x64",
+        ".globl %s\n"
+        "\t.section .rdata,\"dr\"\n"
+        "\t.align 16\n" /* Either align 8 bytes or 2^8 (256) bytes. 8 bytes is needed. */
+        "%s:\n\n",
+
+        ".long ","",HEX_0X
+    },
     {"sun",
         "\t.section \".rodata\"\n"
         "\t.align   8\n"
